File: 103-refactoring-view-controllers-part-2/CleanTableViews/CleanTableViews/ArrayDataSource.m
//
//  ArrayDataSource.m
//  CleanTableViews
//
//  Created by ben on 1/14/14.
//  Copyright (c) 2014 NSScreencast. All rights reserved.
//

#import "ArrayDataSource.h"

@interface ArrayDataSource ()

@property (nonatomic, copy) NSString *cellIdentifier;
@property (nonatomic, copy) ConfigureCellBlock cellBlock;

@end

@implementation ArrayDataSource

- (id)initWithItems:(NSArray *)items
     cellIdentifier:(NSString *)identifier
 configureCellBlock:(ConfigureCellBlock)configureCellBlock {
    self = [super init];
    if (self) {
        self.items = items;
        self.cellIdentifier = identifier;
        self.cellBlock = configureCellBlock;
    }
    
    return self;
}

- (id)itemAtIndexPath:(NSIndexPath *)indexPath {
    return self.items[indexPath.row];
}

- (NSInteger)tableView:(UITableView*)tableView
 numberOfRowsInSection:(NSInteger)section {
    return self.items.count;
}

- (UITableViewCell*)tableView:(UITableView*)tableView
        cellForRowAtIndexPath:(NSIndexPath*)indexPath {
    id cell = [tableView dequeueReusableCellWithIdentifier:self.cellIdentifier
                                              forIndexPath:indexPath];
    id item = [self itemAtIndexPath:indexPath];
    self.cellBlock(cell,item);
    return cell;
}
@end

File: 274-isolating-cloudkit-from-your-controllers/SharedNotes/SharedNotes/ArrayDatasource.swift
//
//  ArrayDatasource.swift
//  SharedNotes
//
//  Created by Ben Scheirman on 6/1/17.
//  Copyright © 2017 NSScreencast. All rights reserved.
//

import Foundation

class ArrayDatasource<T> {
    
    weak var delegate: DatasourceDelegate?
    
    var objects: Array<T> = []
    
    func object(at indexPath: IndexPath) -> T {
        return objects[indexPath.row]
    }
    
    @discardableResult
    func insert(object: T, at index: Int) -> IndexPath {
        objects.insert(object, at: index)
        let indexPath = IndexPath(item: index, section: 0)
        DispatchQueue.main.async {
            self.delegate?.objectAdded(at: indexPath)
        }
        return indexPath
    }
    
    @discardableResult
    func add(object: T) -> IndexPath {
        objects.append(object)
        let indexPath = IndexPath(item: objects.count - 1, section: 0)
        DispatchQueue.main.async {
            self.delegate?.objectAdded(at: indexPath)
        }
        return indexPath
    }
    
    @discardableResult
    func delete(at index: Int) -> IndexPath {
        objects.remove(at: index)
        let indexPath = IndexPath(row: index, section: 0)
        DispatchQueue.main.async {
            self.delegate?.objectRemoved(at: indexPath)
        }
        return indexPath
    }
}
